### ISSUE TITLE
New value type (static) is not in annotated types.

### ISSUE THREAD

**Issue by donquixote** [kalessil/phpinspectionsea#439]
**steps to reproduce**

1. Create PHP file
```php
<?php

namespace Acme;

interface I {}

class C implements I {

  private $color = 'grey';

  /**
   * @param string $color
   *
   * @return static
   *   A clone with the specified color.
   */
  public function withColor($color) {
    $clone = clone $this;
    $clone->color = $color;
    return $clone;
  }
}

function foo(I $obj) {
  
  if ($obj instanceof C) {
    // "New value type (static) is not in annotated types."
    $obj = $obj->withColor('blue');
  }
  
  return $obj;
}
```

**Expected**
No inspection warning.

**Actual**
Inspection warning: "New value type (static) is not in annotated types." on `$obj = $obj->..` in `foo()`.

From EA Extended > Architecture > Callable parameter usage violates definition.

## Background

I really like this inspection, and would prefer not to disable it.
However, in this case it seems to be not justified. Right?

**Comment by kalessil** [comment]
@donquixote, thank you for reporting.

Yes, static/self/$this are something we didn't focus originally, so it's time to enhance the inspection =)

**Comment by kalessil** [comment]
Fixed, fresh binary oushed to the master branch as well.

**Comment by 3DFace** [comment]
Hi!

Another case with a very similar issue. Should I create a separate report?

```php
<?php

class Animal {

	/**
	 * @return static
	 */
	public static function create() {
		return new static();
	}

}

class Cat extends Animal {
}


class Example {

	/**
	 * @param Cat|null $cat
	 */
	public function acceptCat(?Cat $cat) : void{
		$cat = $cat ?? Cat::create(); // "New value type (static) is not in annotated types"
		echo $cat !== null;
	}

}

```
PhpStorm 2017.3.3 EAP
Build #PS-173.4301.15, built on January 10, 2018

**Comment by kalessil** [comment]
@3DFace : yes, please file a new issue.

**Comment by donquixote** [comment]
I suppose the other issue is this one, #746.
(just adding this for reference)

**Comment by donquixote** [comment]
@kalessil In which version of the plugin is this fixed?
I am using version 3.0.3, which I guess is quite recent?
Maybe the fix only exists in 2.x branch, not in 3.x branch?

**Comment by donquixote** [comment]
Looking at
https://github.com/kalessil/phpinspectionsea/commit/5894e67e8edd23b328ece38edc6ac42db1cb269c
https://github.com/kalessil/phpinspectionsea/blob/C-3.0.3/src/main/java/com/kalessil/phpStorm/phpInspectionsEA/inspectors/semanticalAnalysis/CallableParameterUseCaseInTypeContextInspection.java
I would think that the fix is already in 3.0.3.

Which makes me wonder why I still see this behavior.

**Comment by donquixote** [comment]
@kalessil I think you need to reopen this.
I still see this problem with 3.0.3 and the code example above.

PhpStorm 2018.2 EAP
Build #PS-182.3208.33, built on June 13, 2018
PhpStorm EAP User
Expiration date: July 13, 2018
JRE: 1.8.0_152-release-1226-b7 amd64
JVM: OpenJDK 64-Bit Server VM by JetBrains s.r.o
Linux 4.13.0-43-generic

**Comment by kalessil** [comment]
@donquixote : in 3.0.3 this suppose to be fixed, please file a new issue please with steps to reproduce.

I'll investigate and try to reproduce yours case (we have tests, so quite sure that's a new case or due to EAP).
